---
title: "GVD_pipeline"
author: "Theresa Alexander"
date: "1/9/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Load iDA
```{r}
source("~/Desktop/iDA_files/iDA/R/iDA.R")
source("~/Desktop/iDA_files/iDA/R/LDA.R")
```


#get Geuvadis (GVD) expression and phenotype data
```{r}
GVD <- read.table("~/Downloads/vst_counts_462unduped.txt")

#pheno data
unduped_samples = read.csv("~/Downloads/sampleids.csv")
coldata <- read.csv("~/Downloads/E-GEUV-3.sdrf.csv",  header = TRUE, row.names = "AssayName")
coldata <- coldata[unduped_samples$x,]
coldata_pop_lab <- coldata[,c(2,6)]
```


#Run iDA on GVD data
```{r}
#run iDA
GVD_iDA <- iDA(as.matrix(GVD), set.seed = 0, diag = TRUE)

#c(.18398, .07142, .38961, ..5909, .59307, 1)
#save iDA embedding and idents
GVD_iDA_ident = GVD_iDA[[1]]
GVD_iDAembedding = GVD_iDA[[2]]
rownames(GVD_iDAembedding) = colnames(GVD)

GVD_iDAembedding = cbind(GVD_iDAembedding, coldata_pop_lab[ rownames(GVD_iDAembedding),])


```


#PCA on GVD 
```{r}
#find same variable features
var.features <- iDA::VariableGenes(GVD, mean.low.cutoff = 0.1, mean.high.cutoff = 50, dispersion.cutoff = 1)

#embedding
GVD_PCAembedding = prcomp(GVD[var.features,])
GVD_PCAembedding$sdev

plot(GVD_PCAembedding$sdev[1:20], pch = 19, xlab = "PC", ylab="Standard Deviation", main = "Geuvadis Elbow Plot")
```



#cumulative F-Statistics and R^2 for iDA and PCA embeddings
##iDA F-stat and R^2
```{r}

#getting RSS for iDA across dimensions
iDARSS = c()
for (i in 1:(dim(GVD_iDAembedding)[2]-2)) {
  iDARSS[i] = anova(lm(as.numeric(GVD_iDAembedding[,i]) ~ as.factor(coldata$Population)*as.factor(coldata$laboratory)))[4,2]
}


#getting TSS for iDA across dimensions

iDATSS = c()
for (i in 1:(dim(GVD_iDAembedding)[2]-2)) {
  iDATSS[i] = sum(anova(lm(as.numeric(GVD_iDAembedding[,i]) ~ as.factor(coldata$Population)*as.factor(coldata$laboratory)))[,2])
}




Fstat_GVD_iDA_maineffects = c((iDATSS[1]-iDARSS[1])/iDARSS[1], 
                       (sum(iDATSS[1:2])-sum(iDARSS[1:2]))/sum(iDARSS[1:2]), 
                       (sum(iDATSS[1:3])-sum(iDARSS[1:3]))/sum(iDARSS[1:3]), 
                       (sum(iDATSS[1:4])-sum(iDARSS[1:4]))/sum(iDARSS[1:4]), 
                       (sum(iDATSS[1:5])-sum(iDARSS[1:5]))/sum(iDARSS[1:5]), 
                       (sum(iDATSS[1:6])-sum(iDARSS[1:6]))/sum(iDARSS[1:6]), 
                       (sum(iDATSS[1:7])-sum(iDARSS[1:7]))/sum(iDARSS[1:7]))



R2_GVD_iDA_maineffects = c(1-(iDARSS[1]/iDATSS[1]), 
                       1-sum(iDARSS[1:2])/sum(iDATSS[1:2]), 
                       1-sum(iDARSS[1:3])/sum(iDATSS[1:3]), 
                       1-sum(iDARSS[1:4])/sum(iDATSS[1:4]), 
                       1-sum(iDARSS[1:5])/sum(iDATSS[1:5]), 
                       1-sum(iDARSS[1:6])/sum(iDATSS[1:6]), 
                       1-sum(iDARSS[1:7])/sum(iDATSS[1:7]))

```



##PCA F-stat and R^2
```{r}


#getting RSS for PCA across dimensions
PCRSS = c()
for (i in 1:(dim(GVD_iDAembedding)[2]-2)) {
  PCRSS[i] = anova(lm(as.numeric(GVD_PCAembedding[,i]) ~ as.factor(coldata$Population)*as.factor(coldata$laboratory)))[4,2]
}



#getting TSS for PCA across dimensions
PCTSS = c()
for (i in 1:(dim(GVD_iDAembedding)[2]-2)) {
  PCTSS[i] = sum(anova(lm(as.numeric(GVD_PCAembedding[,i]) ~ as.factor(coldata$Population)*as.factor(coldata$laboratory)))[,2])
}


#compute cumulative F-statistic for each dimension
Fstat_GVD_PCA_maineffects = c((PCTSS[1]-PCRSS[1])/PCRSS[1], 
                       (sum(PCTSS[1:2])-sum(PCRSS[1:2]))/sum(PCRSS[1:2]), 
                       (sum(PCTSS[1:3])-sum(PCRSS[1:3]))/sum(PCRSS[1:3]), 
                       (sum(PCTSS[1:4])-sum(PCRSS[1:4]))/sum(PCRSS[1:4]), 
                       (sum(PCTSS[1:5])-sum(PCRSS[1:5]))/sum(PCRSS[1:5]), 
                       (sum(PCTSS[1:6])-sum(PCRSS[1:6]))/sum(PCRSS[1:6]), 
                       (sum(PCTSS[1:7])-sum(PCRSS[1:7]))/sum(PCRSS[1:7]))


#compute cumulative R^2 for each dimension
R2_GVD_PCA_maineffects = c(1-(PCRSS[1]/PCTSS[1]), 
                       1-sum(PCRSS[1:2])/sum(PCTSS[1:2]), 
                       1-sum(PCRSS[1:3])/sum(PCTSS[1:3]), 
                       1-sum(PCRSS[1:4])/sum(PCTSS[1:4]), 
                       1-sum(PCRSS[1:5])/sum(PCTSS[1:5]), 
                       1-sum(PCRSS[1:6])/sum(PCTSS[1:6]), 
                       1-sum(PCRSS[1:7])/sum(PCTSS[1:7]))




Fstat_GVD = as.data.frame(cbind(Fstat_GVD_iDA_maineffects, Fstat_GVD_PCA_maineffects))
R2_GVD = as.data.frame(cbind(R2_GVD_iDA_maineffects, R2_GVD_PCA_maineffects))


```


##Plot F-stat and R^2
```{r}
ggplot(Fstat_GVD, aes(x = as.factor(1:7))) + geom_point(aes(y = Fstat_GVD$Fstat_GVD_iDA_maineffects,color = "iDA")) + geom_line(aes(x = 1:7, y = Fstat_GVD$Fstat_GVD_iDA_maineffects, color = "iDA")) + geom_point(aes(y = Fstat_GVD$Fstat_GVD_PCA_maineffects,color = "PCA")) + geom_line(aes(x = 1:7, y = Fstat_GVD$Fstat_GVD_PCA_maineffects, color = "PCA"))  +  scale_colour_manual("", breaks = c("iDA", "PCA"), values = c("blue", "orange")) + labs( x = "Dimension", y = "Cumulative F-statistic") + theme_classic() 



ggplot(R2_GVD, aes(x = as.factor(1:7))) + geom_point(aes(y = R2_GVD$R2_GVD_iDA_maineffects,color = "iDA")) + geom_line(aes(x = 1:7, y = R2_GVD$R2_GVD_iDA_maineffects, color = "iDA")) + geom_point(aes(y = R2_GVD$R2_GVD_PCA_maineffects,color = "PCA")) + geom_line(aes(x = 1:7, y = R2_GVD$R2_GVD_PCA_maineffects, color = "PCA"))  +  scale_colour_manual("", breaks = c("iDA", "PCA"), values = c("blue", "orange")) + labs( x = "Dimension", y = "Cumulative" ~ R^2) + theme_classic() 


```


#plot LD versus genes in eQTL with YRI and EUR populations
##get SNP file
```{r}
SNPs_eQTL_YRI = read.csv("~/Downloads/YRI_eQTL_SNPs.csv", header = TRUE, stringsAsFactors = FALSE)
```

##rs143415501-RAB5C eQTL pair
```{r}

#rs143415501 AFR MAF = .06, EUR MAF = 0
ENSG00000108774_10_iDA = as.data.frame(cbind(GVD_iDAembedding, t(GVD["ENSG00000108774.10",])))

#plot LD7 versus RAB5C gene expression
ggplot(data = ENSG00000108774_10_iDA, aes(x = LD3, y = ENSG00000108774.10, color = Population), size = 1.2) + 
  geom_point(aes(shape = as.factor(coldata$laboratory)), size = 1.2) + 
  xlab( "LD7 Weight") + 
  ylab("RAB5C Expression") + 
  theme_classic() + 
  labs(color='Population', shape = "Laboratory") + 
  scale_shape_manual(values=c(15, 16, 17, 18, 13, 8, 6)) 


```



##rs11757158-SRF eQTL pair
```{r}
#rs11757158 SNP has AFR MAF = .45, EUR MAF = .09
ENSG000001126586_iDA = as.data.frame(cbind(GVD_iDAembedding, t(GVD["ENSG00000112658.6",])))

#plot LD7 versus SRF gene expression
ggplot(data = ENSG000001126586_iDA, aes(x = ENSG000001126586_iDA$`7`, y = ENSG000001126586_iDA$ENSG00000112658.6, color = ENSG000001126586_iDA$Population), size = 1.2) + 
  geom_point(aes(shape = as.factor(coldata$laboratory)), size = 1.2)  +  
  xlab( "LD7 Weight") + 
  ylab("SRF Expression") + 
  theme_classic() + 
  labs(color='Population', shape = "Laboratory") + 
  scale_shape_manual(values=c(15, 16, 17, 18, 13, 8, 6))



```

##rs114313536-PSKH1 eQTL pair
```{r}
#rs114313536 AFR MAF = .16, EUR MAF = .02
ENSG000001597924_iDA = as.data.frame(cbind(GVD_iDAembedding, t(GVD["ENSG00000159792.4",])))


#plot LD7 versus PSKH1 gene expression
ggplot(data = ENSG000001597924_iDA, aes(x = ENSG000001597924_iDA$`7`, y = ENSG000001597924_iDA$ENSG00000159792.4, color = ENSG000001597924_iDA$Population), size = 1.2) + 
  geom_point(aes(shape = as.factor(coldata$laboratory)), size = 1.2) + 
  xlab( "LD7 Weight") + 
  ylab("PSKH1 Expression") + 
  theme_classic() + 
  labs(color='Population', shape = "Laboratory") + 
  scale_shape_manual(values=c(15, 16, 17, 18, 13, 8, 6))

```


##rs200846953-PWAR6 eQTL pair
```{r}
#rs200846953 AFR MAF = 0.54, EUR MAF = 	0.83
ENSG000002571511_iDA = as.data.frame(cbind(GVD_iDAembedding, t(GVD["ENSG00000257151.1",])))

#plot LD7 versus PWAR6 gene expression
ggplot(data = ENSG000002571511_iDA, aes(x = ENSG000002571511_iDA$`7`, y = ENSG000002571511_iDA$ENSG00000257151.1, color = ENSG000002571511_iDA$Population)) + 
  geom_point(aes(shape = as.factor(coldata$laboratory)), size = 1.2)  +  
  xlab( "LD7 Weight") + 
  ylab("PWAR6 Expression") + 
  theme_classic() + 
  labs(color='Population', shape = "Laboratory") + 
  scale_shape_manual(values=c(15, 16, 17, 18, 13, 8, 6))

```


#Dunn index
```{r}
library(clValid)
samplebygeneGVD = t(GVD)

# distance matrix (takes awhile to compute euclidean distance)
Dist_GVD <- dist(samplebygeneGVD, method="euclidean")

#Given populatino and laboratory clusters
GVD_pop_lab = coldata$poplab
#iDA clusters
iDAGVD_clusters = GVD_iDA_ident

poplabclusts_Dunn = dunn(Dist_GVD, as.numeric(GVD_pop_lab))
iDAclustsGVD_Dunn = dunn(Dist_GVD, as.numeric(iDAGVD_clusters))


poplabclusts_Dunn
iDAclustsGVD_Dunn
```

